### PR TITLE
Upgrade xgboost to 1.5.1 and upgrade spark to 3.2.0

### PR DESCRIPTION
**Pull request by WeichenXu123**
Upgrade xgboost to 1.5.1 and upgrade spark to 3.2.0
See this CI running result:
https://app.travis-ci.com/github/WeichenXu123/mleap/builds/245053199

### CODE REVIEW

File: mleap-databricks-runtime/build.sbt
Comment (WeichenXu123) [inline mleap-databricks-runtime/build.sbt:11]: @wenfeiy-db Is this part sensitive and we should not put in OSS repo ?

File: project/Dependencies.scala
Comment (WeichenXu123) [inline project/Dependencies.scala:17]: the mleap-spark require slf4jVersion >= 1.7.30 if we upgrade spark to 3.2.0. because slf4jVersion new version add new methods which used by spark 3.2

File: mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/XGBoostPredictorClassificationModelParitySpec.scala
Comment (WeichenXu123) [inline mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/XGBoostPredictorClassificationModelParitySpec.scala:104]: The default precision limit (1e-7) will break the test under latest xgboost version.

File: mleap-spark-testkit/src/main/scala/org/apache/spark/ml/parity/SparkParityBase.scala
Comment (WeichenXu123) [inline mleap-spark-testkit/src/main/scala/org/apache/spark/ml/parity/SparkParityBase.scala:78]: Use a global spark context across all suites. Previous test shutdown spark context when every suite finished, but the spark context variable are referred in many places which cause error like: `SparkContext has been shutdown`
In mleap spark related tests, all tests using the same config spark context, so we can use a global one instead.

File: mleap-xgboost-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/mleap/XGBoostClassificationModelParitySpec.scala
Comment (WeichenXu123) [inline mleap-xgboost-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/mleap/XGBoostClassificationModelParitySpec.scala:31]: The xgboost estimator require param dict key using format like "xx_yy_zz", not "xxYyZz", so the master test all these params with capital letters are not set correctly.

I rewrite these param keys as format "xx_yy_zz". Note I don't add some params (e.g. max_depth) because those param don't be saved by model serializer so it will break the `serializes/deserializes the Spark model properly` test.  We can improve this in follow-up PR.

File: mleap-xgboost-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/mleap/XGBoostRegressionModelParitySpec.scala
Comment (WeichenXu123) [inline mleap-xgboost-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/mleap/XGBoostRegressionModelParitySpec.scala:21]: ditto.

**Comment by WeichenXu123** [comment]
CI passed:
https://app.travis-ci.com/github/WeichenXu123/mleap/builds/245053199
(The official repo CI is disabled, I use my own repo CI to run it.)

**Comment by WeichenXu123** [comment]
@ancasarb @jsleight Would you have time to take a look ?

**Comment by ancasarb** [comment]
hi @WeichenXu123 i'll take a look shortly, could I please ask to re-trigger the build (apologies we had run out of free credits on travis-ci and it didn't trigger before)? thank you 🙇‍♀️

File: mleap-spark-testkit/src/main/scala/org/apache/spark/ml/parity/SparkParityBase.scala
Comment (jsleight) [reply mleap-spark-testkit/src/main/scala/org/apache/spark/ml/parity/SparkParityBase.scala:78]: ++ for having a single spark session

File: mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/XGBoostPredictorClassificationModelParitySpec.scala
Comment (jsleight) [reply mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/XGBoostPredictorClassificationModelParitySpec.scala:104]: This makes me a little nervous.  Perhaps we also need to upgrade xgboost predictor library?

**Comment by WeichenXu123** [comment]
@ancasarb Test passed. Could you take a look ? Thanks!

**Comment by ancasarb** [comment]
@WeichenXu123 thanks, changes look fine to me. there are two questions from @jsleight could you please take a look? happy to merge this in once those are addressed, thanks again!

File: mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/XGBoostPredictorClassificationModelParitySpec.scala
Comment (WeichenXu123) [reply mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/XGBoostPredictorClassificationModelParitySpec.scala:104]: xgboost predictor library is already the newest version.
 the precision=1e-5 i think it is fine. New version xgboost algorithm might change slightly so it affect the precision but very small difference.

File: mleap-xgboost-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/mleap/XGBoostClassificationModelParitySpec.scala
Comment (WeichenXu123) [reply mleap-xgboost-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/mleap/XGBoostClassificationModelParitySpec.scala:31]: I don't check detail code, but in my test, the camel style does not work, the parameter does not set correctly.
e.g. if we set `xgboostParams = Map("maxDepth" -> 2)` then the generated estimator the "maxDepth" is still 6 (default value) not 2.

File: mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/XGBoostPredictorClassificationModelParitySpec.scala
Comment (jsleight) [reply mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/XGBoostPredictorClassificationModelParitySpec.scala:104]: I guess the question is more whether the xgboost predictor library itself needs some PR to upgrade it for xgboost 1.5.x support?

Agreed that some change to xgboost might have changed it's decisions a little bit, but the model predictions still should be reproducible and change in the same way.  These tests are just ensuring that the pre- and post-serialized models give the same results.

I tend to get nervous about these xgboost precision details because I have previously encountered issues where little tiny differences in the xgboost setup/usage can impact the predictions in major ways (e.g., whether you use 32 or 64 bit floats).  However, since our testing model/data are simple they might only demonstrate small changes to output, while a more real-world model/data would demonstrate predictions being completely broken.

File: mleap-xgboost-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/mleap/XGBoostClassificationModelParitySpec.scala
Comment (jsleight) [reply mleap-xgboost-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/mleap/XGBoostClassificationModelParitySpec.scala:31]: that's weird.  You're right that the xgboost scala tests are using snake case though.  So ok to leave it!

File: mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/XGBoostPredictorClassificationModelParitySpec.scala
Comment (WeichenXu123) [reply mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/XGBoostPredictorClassificationModelParitySpec.scala:104]: @jsleight 
I test again. It turns out seems previously I test against wrong predictor version.
Current PR when I increase prediction tolerance to 1e-7 it still pass test.

Comment (WeichenXu123) [reply mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/XGBoostPredictorClassificationModelParitySpec.scala]: I revert the change here.

**Review by ancasarb — APPROVED** [review]
lgtm, will merge it now.

File: mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/XGBoostPredictorClassificationModelParitySpec.scala
Comment (jsleight) [reply mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/XGBoostPredictorClassificationModelParitySpec.scala]: awesome!  Thanks for reconfirming ❤️

**Comment by indranilr** [comment]
Gents , is this upgrade due for release anytime soon ? Would appreciate your inputs on a ballpark timeline.

**Comment by jsleight** [comment]
Releasing mleap v0.20.0 seems justified to me.  We've upgraded spark, xgboost, and tensorflow to latest versions.  @ancasarb usually handles the release process.

@indranilr if you urgently need the spark 3.2.0 support (or xgboost 1.5) then you can use the [snapshots from sonatype](https://oss.sonatype.org/content/repositories/snapshots/ml/combust/mleap/) which are always based on current master branch.

**Comment by indranilr** [comment]
> Releasing mleap v0.20.0 seems justified to me. We've upgraded spark, xgboost, and tensorflow to latest versions. @ancasarb usually handles the release process.
> 
> @indranilr if you urgently need the spark 3.2.0 support (or xgboost 1.5) then you can use the [snapshots from sonatype](https://oss.sonatype.org/content/repositories/snapshots/ml/combust/mleap/) which are always based on current master branch.

Thanks for your response, I will have our team evaluate the snapshots till the new release happens, hopefully 0.20 release will be available by end of next month (Correct me if its too early to assume availability. its just that our product release team requires a official release of dependencies to be used)

**Comment by ancasarb** [comment]
hey @indranilr @jsleight, just a quick note to say that i released a new version (0.20) today, it should be available in maven central shortly.

**Comment by indranilr** [comment]
@ancasarb thanks much, we will use this version for our next product release cycle.

**Pull request merged**
